lib/src/main/java/io/ably/lib/http/Http.java

```diff
@@ -222,35 +262,35 @@ private static AblyException checkFor(Throwable t) {
 	}
 
 	<T> T ablyHttpExecute(String path, String method, Param[] headers, Param[] params, RequestBody requestBody, ResponseHandler<T> responseHandler) throws AblyException {
-		try {
-			URL url = buildURL(scheme, getPrefHost(), path, params);
-			return httpExecute(url, method, headers, requestBody, true, responseHandler);
-		} catch(HostFailedException bhe) {
-			/* one of the exceptions occurred that signifies a problem reaching the host */
-			String[] fallbackHosts = Defaults.getFallbackHosts(ably.options);
-			if(fallbackHosts != null) {
-				for(String host : fallbackHosts) {
-					try {
-						URL url = buildURL(scheme, host, path, params);
-						return httpExecute(url, method, headers, requestBody, true, responseHandler);
-					} catch(HostFailedException bhe2) {}
+
+		int retryCountRemaining = isFallbackSupported()?options.httpMaxRetryCount:0;
+		String hostCurrent = this.host;
+		URL url;
+
+		do {
+			url = buildURL(scheme, hostCurrent, path, params);
+
+			try {
+				return httpExecute(url, method, headers, requestBody, true, responseHandler);
+			} catch (HostFailedException e) {
+				retryCountRemaining--;
+
+				if (retryCountRemaining >= 0) {
+					Log.d(TAG, "Connection failed to host `" + hostCurrent + "`. Searching for new host...");
+					hostCurrent = Hosts.getFallback(hostCurrent);
+					Log.d(TAG, "Switched to `" + hostCurrent + "`.");
 				}
 			}
-			throw new AblyException("Connection failed; no host available", 404, 80000);
-		}
+		} while (retryCountRemaining >= 0);
+
+		throw new AblyException("Connection failed; no host available", 404, 80000);
```

Comment:
I am not sure why we would throw this generic exception instead of re-raising the last exception failure?